backend/rust/burn/src/main.rs

```diff
@@ -0,0 +1,89 @@
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::transport::Server;
+use tonic::{async_trait, Request, Response, Status};
+use tracing;
+use tracing_subscriber;
+use bunker::pb::{EmbeddingResult, GenerateImageRequest, HealthMessage, ModelOptions, PredictOptions, Reply, StatusResponse, TokenizationResponse, TranscriptRequest, TranscriptResult, TtsRequest};
+use bunker::pb::Result as PbResult;
+use bunker::service::BackendService;
+
+
+// implement BackendService trait in bunker
+
+struct BurnBackend;
+
+#[async_trait]
+impl BackendService<ReceiverStream<Result<Reply, Status>>> for BurnBackend{
+
+    async fn health(&self, request: Request<HealthMessage>) -> Result<Response<Reply>,Status> {
+        // return a Result<Response<Reply>,Status>
+        let reply = Reply {
+            message: "OK".into(),
+        };
+        let res=Response::new(reply);
+        Ok(res)
+    }
+
+    async fn predict(&self, request: Request<PredictOptions>) -> Result<Response<Reply>,Status> {
+        todo!()
+    }
+
+    async fn load_model(&self, request: Request<ModelOptions>) -> Result<Response<PbResult>, Status> {
+        todo!()
+    }
+
+    async fn predict_stream(&self, request: Request<PredictOptions>) -> Result<Response<ReceiverStream<Result<Reply,Status>>>, Status> {
+        todo!()
+    }
+
+    async fn embedding(&self, request: Request<PredictOptions>) -> Result<Response<EmbeddingResult>, Status> {
+        todo!()
+    }
+
+    async fn generate_image(&self, request: Request<GenerateImageRequest>) -> Result<Response<PbResult>, Status> {
+        todo!()
+    }
+
+    async fn audio_transcription(&self, request: Request<TranscriptRequest>) -> Result<Response<TranscriptResult>, Status> {
+        todo!()
+    }
+
+    async fn tts(&self, request: Request<TtsRequest>) -> Result<Response<PbResult>, Status> {
+        todo!()
+    }
+
+    async fn tokenize_string(&self, request: Request<PredictOptions>) -> Result<Response<TokenizationResponse>, Status> {
+        todo!()
+    }
+
+    async fn status(&self, request: Request<HealthMessage>) -> Result<Response<StatusResponse>, Status> {
+        todo!()
+    }
+
+}
+
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+
+    let subscriber = tracing_subscriber::fmt()
+        .compact()
+        .with_file(true)
+        .with_line_number(true)
+        .with_target(true)
+        .with_level(true)
+        .finish();
+
+    tracing::subscriber::set_global_default(subscriber)
+        .expect("setting default subscriber failed");
+
+    let addr = "[::1]:50052".parse().unwrap();
+
+    let backend = BackendService {};
+
+    let svc = BurnBackend::new(backend);
+
+    Server::builder().add_service(svc).serve(addr).await?;
```

Comment:
Ideally you can move even the prost-specific instantiation in the library crate so you depend only on bunker.

Reply:
Please give an example here. I am trying to implement this in the `bunker/src/service.rs` under the trait. But I got some warning about the unsafe thread of `BackendService`.

Reply:
I am trying to implement a function in trait like, but many of issues of it.

```    async fn start(&self) -> Result<(), Box<dyn std::error::Error>> {
        let addr = "[::1]:50051".parse().unwrap();
        let backend = self.clone();
        let backend_safe = Arc::new(Mutex::new(backend));
        let svc = BackendServer::new(backend_safe);

        tokio::spawn(async move {
            Server::builder()
                .add_service(svc)
                .serve(addr)
                .await
                .unwrap();
        });

        Ok(())
    }
```

Reply:
Why spawning it in a separate task?

Reply:
Maybe this one will be good to do it:

```rust

pub trait BackendService<T = ReceiverStream<Result<Reply, Status>>>{};

impl<T, Caller> BackendService <T> for Caller {
    fn startup(self, _:T){}
}
```

Reply:
I'd first with an async function like the one I put as example, just to keep the burn backend simple, then iterate to make more structured as we see how it feels.

Reply:
Thank you.

Reply:
> Why spawning it in a separate task?

I am trying to solve the `unsafe thread` issue but it is make no sense. I already removed it.